Detectors/MUON/MCH/Raw/Decoder/src/DataDecoder.cxx

```diff
@@ -203,18 +263,113 @@ void DataDecoder::decodeBuffer(gsl::span<const std::byte> page)
   if (!mDecoder) {
     DecodedDataHandlers handlers;
     handlers.sampaChannelHandler = channelHandler;
+    handlers.sampaHeartBeatHandler = heartBeatHandler;
     mDecoder = mFee2Solar ? o2::mch::raw::createPageDecoder(page, handlers, mFee2Solar)
                           : o2::mch::raw::createPageDecoder(page, handlers);
   }
   mDecoder(page);
 
   if (mDebug) {
+    std::cout << "[decodeBuffer] mOrbits size: " << mOrbits.size() << std::endl;
+    //dumpOrbits();
     std::cout << "[decodeBuffer] mOutputDigits size: " << mOutputDigits.size() << std::endl;
     dumpDigits(true);
     dumpDigits(false);
   }
 };
 
+int32_t digitsTimeDiff(uint32_t orbit1, uint32_t bc1, uint32_t orbit2, uint32_t bc2)
+{
+  // bunch crossings are stored with 20 bits
+  static const int32_t BCROLLOVER = (1 << 20);
+  // bunch crossings half range
+  //static const int32_t BCHALFRANGE = (1 << 19);
+  // number of bunch crossings in one orbit
+  static const int32_t BCINORBIT = 3564;
+
+  // We use the difference of orbits values to estimate the minimum and maximum allowed
+  // difference in bunch crossings
+  int64_t dOrbit = static_cast<int64_t>(orbit2) - static_cast<int64_t>(orbit1);
+
+  // Digits might be sent out later than the orbit in which they were recorded.
+  // We account for this by allowing an extra +/- 2 orbits when converting the
+  // difference from orbit numbers to bunch crossings.
+  int64_t dBcMin = (dOrbit - 2) * BCINORBIT;
+  int64_t dBcMax = (dOrbit + 2) * BCINORBIT;
+
+  // Difference in bunch crossing values
+  int64_t dBc = static_cast<int64_t>(bc2) - static_cast<int64_t>(bc1);
+
+  if (dBc < dBcMin) {
+    // the difference is too small, so we assume that it needs to be
+    // incremented by one rollover factor
+    dBc += BCROLLOVER;
+  } else if (dBc > dBcMax) {
+    // the difference is too big, so we assume that it needs to be
+    // decremented by one rollover factor
+    dBc -= BCROLLOVER;
+  }
+
+  return static_cast<int32_t>(dBc);
+}
+
+void DataDecoder::computeDigitsTime_(std::vector<o2::mch::Digit>& digits, std::vector<SampaInfo>& sampaInfo, TimeFrameInfos& timeFrameInfos, bool debug)
+{
+  if (digits.size() != sampaInfo.size()) {
```

Comment:
shouldn't this be an error ?

Reply:
yes, but I wasn't sure if to throw an exception, and which exception would be appropriate...